server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala

```diff
@@ -65,6 +82,36 @@ object Email {
   }
 }
 
+object ReadLevel {
+  private val metadataProperty: Seq[NonEmptyString] = Seq("id", "size", "mailboxIds",
+    "mailboxIds", "blobId", "threadId", "receivedAt")
+  private val fullProperty: Seq[NonEmptyString] = Seq("bodyStructure", "textBody", "htmlBody",
+    "attachments", "headers", "bodyValues", "preview", "hasAttachment")
+
+  def of(property: NonEmptyString): ReadLevel = if (metadataProperty.contains(property)) {
+    MetadataReadLevel
+  } else if (fullProperty.contains(property)) {
+    FullReadLevel
+  } else {
+    HeaderReadLevel
+  }
+
+  def combine(readLevel1: ReadLevel, readLevel2: ReadLevel): ReadLevel = readLevel1 match {
+    case MetadataReadLevel => readLevel2
+    case FullReadLevel => FullReadLevel
+    case HeaderReadLevel => readLevel2 match {
+      case MetadataReadLevel => HeaderReadLevel
```

Comment:
you don't need this nested pattern matching if you return 'HeaderReadLevel' for each possible values

Reply:
Obviosly we should not :-)